src/HttpRequest.cc

```diff
@@ -739,3 +739,60 @@ HttpRequest::manager(const CbcPointer<ConnStateData> &aMgr, const AccessLogEntry
     }
 }
 
+/// a helper for validating FindListeningPortAddress()-found address candidates
+static const Ip::Address *
+FindListeningPortAddressInAddress(const Ip::Address *ip)
+{
+    // FindListeningPortAddress() callers do not want INADDR_ANY addresses
+    return (ip && !ip->isAnyAddr()) ? ip : nullptr;
+}
+
+/// a helper for handling PortCfg cases of FindListeningPortAddress()
+static const Ip::Address *
+FindListeningPortAddressInPort(const AnyP::PortCfgPointer &port)
+{
+    return port ? FindListeningPortAddressInAddress(&port->s) : nullptr;
+}
+
+/// a helper for handling Connection cases of FindListeningPortAddress()
+static const Ip::Address *
+FindListeningPortAddressInConn(const Comm::ConnectionPointer &conn)
+{
+    return conn ? FindListeningPortAddressInAddress(&conn->local) : nullptr;
+}
+
+const Ip::Address *
+FindListeningPortAddress(const HttpRequest *callerRequest, const AccessLogEntry *ale)
+{
+    // Check all sources of usable listening port information, giving
+    // HttpRequest and masterXaction a preference over ALE.
+
+    const HttpRequest *request = callerRequest;
+    if (!request && ale)
+        request = ale->request;
+    if (!request)
+        return nullptr; // not enough information
+
+    const Ip::Address *ip = nullptr;
+
+    if (request->flags.interceptTproxy || request->flags.intercepted) {
+        ip = FindListeningPortAddressInPort(request->masterXaction->squidPort);
+        if (!ip && ale)
+            ip = FindListeningPortAddressInPort(ale->cache.port);
+        if (ip)
+            return ip;
+
+        // XXX: tcpClient contains client destination rather than Squid
```

Comment:
@rousskov ; I disagree with the assesment that this PR is not adding a regression with this XXX. The original code was

     const bool interceptedAtKnownPort = al->request ?
                                                (al->request->flags.interceptTproxy ||
                                                 al->request->flags.intercepted) && al->cache.port :
                                                false;
            if (interceptedAtKnownPort) {
                ...
                    out = al->cache.port->s.toStr(tmp, sizeof(tmp));
            } else if (al->tcpClient)

Only the cases where there was no request AND no squidPort would intercepted traffic fall through to the tcpClient use. Any case where request and squidPort were available would skip past the else block where tcpClient was being used.

With the new logic being more robust in where it can find the squidPort object I think we should "return ip;" (even if its nil) where this XXX is written. Otherwise we would suddenly start producing tcpClient address in many cases where ADDR_ANY was configured and currently logging "-" correctly. Only the currently buggy case (known to be intercepted but no al->cache.port available) would change and that would be a good change toward better accuracy.

Reply:
> I disagree with the assesment that this PR is not adding a regression with this XXX

Agreed. My summary statement was wrong. This PR is currently increasing the number of potentially buggy cases. It also increases the number of potentially improved cases. Since I was not particularly interested in spending time on figuring out which buggy mess is better, I deleted my negative vote. That removal does not mean I am excited about the current PR code -- that XXX is there for a good reason, and more problems have been uncovered since that revision. It only means I trust you and the author to make the right call.

> I think we should "return ip;" (even if its nil) where this XXX is written

Yes, that was my recommendation as well. I would go one or even two steps further:

1. `return ip` for intercepted callers (we agree here)
2. remove the interception condition (use port sources first, for all callers)
3. remove tcpClient code (use only port sources, for all callers)

Reply:
Well in this commit, I did raise the question if falling back to localIp ok? (with XXX note)

https://github.com/squid-cache/squid/pull/198/commits/360ff9c7d2348559442deb864ebcbba10ecfbe70

> + // XXX: is it correct to return localIp if cacheIp didn't have info? 

And then later @rousskov in his commit mentioned that

> 3) 8095d23 -- resolves two out of three PR XXXs by consistently going after all available source of information and checking isAnyAddr(). Please see the commit message for details.

He changed my XXX wording with his. So I thought he is aware of this change and he is ok with it.

Also in my comment yesterday here: https://github.com/squid-cache/squid/pull/198#issuecomment-389114891

I did infact mention about returning nullptr and not return tcpClient. I wrote "(more or less)" specifically because there was slight change in original logic. (the one Amos is now objecting to :) )

> We can decide on OK for now. And keep existing behaviour (more or less).
> 
> Later we can add "return nullptr;" at end of the "interception" logic and not allow it to go ahead and return tcpClient.